### ISSUE TITLE
tailscale 1.20.1 with exit node in container stops working after ~1 hour

### ISSUE THREAD

**Issue by razielgn** [tailscale/tailscale#3745]
### What is the issue?

What happened: socks5 proxy via exit-node stops working after some time.
What I expected: socks5 proxy keeps working indefinetely. 

### Steps to reproduce

1. tailscale is launched in a container, here you can see the `Dockerfile` and the `entrypoint.sh`: https://gist.github.com/razielgn/b2c538ffe6f340e23c53c77c45aef336
2. another container connects to it via socks5

After sometime, somewhat 1 hour, maybe a bit less, the socks5 proxy stops answering.
There is a healthcheck in place on the container, using `curl`.
When it stops working, it reports:

```
curl: (7) Failed to connect to 0.0.0.0 port 1055 after 0 ms: Connection refused
```

With that, tailscale (debug logging) stops logging. Here you can find the full logging output https://gist.github.com/razielgn/98458ff8edfac897959bc6968321f115

I'll leave the container running if you want me run something like `strace` to provide further data.

### Are there any recent changes that introduced the issue?

Not sure, as before 1.20.0, socks5 did not support exit nodes.

### OS

Linux

### OS version

5.15.13-1-lts

### Tailscale version

1.20.1

### Bug report

BUG-ff70fb09d6d5cd901398e9a21c1f4b3561bd461d6888e8956dfa7f4b9c3e036b-20220115210524Z-c49f1036e5cf7312

**Comment by bradfitz** [comment]
> another container connects to it via socks5

I don't follow. Socks5 is for outbound connections. That's for letting programs running on the same machine as the tailscaled in userspace-networking mode to be able to talk to other peers.

Concretely, which are your two Tailscale IPs and which is running the socks5 server and which is/are in userspace-networking mode?

**Comment by razielgn** [comment]
> > another container connects to it via socks5
> 
> I don't follow. Socks5 is for outbound connections. That's for letting programs running on the same machine as the tailscaled in userspace-networking mode to be able to talk to other peers.
> 
> Concretely, which are your two Tailscale IPs and which is running the socks5 server and which is/are in userspace-networking mode?

I'm using the userspace tailscaled as socks5 proxy for talking to other tailscale nodes and the rest of the traffic exits from the selected exit node.

```
[container 1 with curl] ==socks5==> [container 2 with tailscaled] ==tailscale conn==> [exit node] ==> [internet]
```

Maybe I misunderstood such functionality and its intended usage, but up until tailscaled in container 2 stops working, I get the correct behaviour, which is routing requests from container 1 through the exit node.

**Comment by bradfitz** [comment]
So container 2 (the userspace-networking tailscaled running the SOCKS5 server) doesn't crash after an hour but just stops working?

Can you run `tailscale debug daemon-goroutines` on it and attach its output with all its goroutines?

Also, `lsof -p $(pidof tailscaled)` to see whether it's leaking file descriptors or something.

**Comment by bradfitz** [comment]
If it's leaking fds in SOCKS5, maybe the health check is contributing to its demise by making them pile up at a predictable-ish rate and that's what adds up to an hour.

**Comment by bradfitz** [comment]
@razielgn, is it crashing due to OOM? Do you see the first tailscaled's memory grow and grow and then get killed by the kernel?

You might be hitting #3762

**Comment by bradfitz** [comment]
This is now fixed in unstable build 1.21.21 and will be in the stable build 1.20.2 later today.

**Comment by razielgn** [comment]
Thank you!
